addons/binding/org.openhab.binding.sleepiq/src/main/java/org/openhab/binding/sleepiq/handler/SleepIQCloudHandler.java

```diff
@@ -0,0 +1,316 @@
+        cloud = SleepIQ.create(cloudConfig);
+
+        logger.info("Authenticating to SleepIQ cloud service");
+        cloud.login();
+
+        logger.info("Authentication successful");
+    }
+
+    @Override
+    public synchronized void dispose() {
+
+        logger.debug("Disposing SleepIQ cloud handler");
+
+        if (pollingJob != null && !pollingJob.isCancelled()) {
+            pollingJob.cancel(true);
+            pollingJob = null;
+        }
+
+        cloud = null;
+    }
+
+    /**
+     * Get the cloud service API.
+     *
+     * @return the service object (may be <code>null</code>)
+     */
+    private synchronized SleepIQ getCloud() {
+        return cloud;
+    }
+
+    /**
+     * Start or stop a background polling job to look for bed status updates based on the whether or not there are any
+     * listeners to notify.
+     */
+    private synchronized void updateListenerManagement() {
+
+        if (cloud == null) {
+            return;
+        }
+
+        if (!bedStatusListeners.isEmpty() && (pollingJob == null || pollingJob.isCancelled())) {
+            int pollingInterval = getConfigAs(SleepIQCloudConfiguration.class).pollingInterval;
+            pollingJob = scheduler.scheduleAtFixedRate(pollingRunnable, pollingInterval, pollingInterval,
+                    TimeUnit.SECONDS);
+        } else if (bedStatusListeners.isEmpty() && pollingJob != null && !pollingJob.isCancelled()) {
+            pollingJob.cancel(true);
+            pollingJob = null;
+        }
+    }
+
+    /**
+     * Retrieve the latest status on all beds and update all registered listeners.
+     */
+    private void publishBedStatusUpdates() {
+        publishBedStatusUpdates(null);
+    }
+
+    /**
+     * Retrieve the latest status on all beds and update the given listener. If the listener is <code>null</code>, all
+     * registered listeners will be updated.
+     *
+     * @param listener the listener to update (may be <code>null</code>)
+     */
+    private void publishBedStatusUpdates(final BedStatusListener listener) {
+
+        SleepIQ cloud = getCloud();
+        if (cloud == null) {
+            return;
+        }
+
+        FamilyStatus status = cloud.getFamilyStatus();
+        for (BedStatus bedStatus : status.getBeds()) {
+
+            if (listener != null) {
+                listener.onBedStateChanged(cloud, bedStatus);
+            } else {
+                bedStatusListeners.stream().forEach(l -> l.onBedStateChanged(cloud, bedStatus));
+            }
+        }
+    }
+
+    /**
+     * Register the given listener to receive bed status updates.
+     *
+     * @param listener the listener to register
+     */
+    public void registerBedStatusListener(final BedStatusListener listener) {
+
+        if (listener == null) {
+            return;
+        }
+
+        bedStatusListeners.add(listener);
+        publishBedStatusUpdates(listener);
```

Comment:
This would work for the first listener because that would trigger the polling job to start and perform an immediate update. However, if I add a new listener (add a second bed to the network), I have to wait up to the polling interval (worst case) before the new bed Thing gets any status. By implementing the way I have, each bed Thing gets its initial status immediately when added. 

This was mainly done for the PaperUI user that is expecting the system to work automatically. Let's assume a scenario where the user has two beds and I am not publishing updates immediately, but rather using an initial schedule delay of 0. The cloud service is connected and the system adds two beds to the Inbox. The user first clicks to accept one of the beds and create a Thing. This starts the polling job immediately and gives a status update to the bed. The user moves on to accept the second bed. This will create a Thing and register the listener, but since the job is already running it will provide no update until the next minute (default). If the user immediately clicks to view the two new beds, he/she will see one with status and one without.

My approach was to push out an update each time a listener is added. The first listener added will start the polling job, but since I have just pushed an update, delay the first execution of the job for the polling interval. Any subsequent listeners that are added will get an immediate update regardless of where the job is at in its polling period and PaperUI users will get immediate feedback.